alacritty_terminal/src/input.rs

```diff
@@ -389,23 +389,32 @@ enum MousePosition {
 }
 
 impl<'a, A: ActionContext + 'a> Processor<'a, A> {
-    fn mouse_position(&mut self, point: Point) -> MousePosition {
+    fn mouse_position(&mut self, point: Point, modifiers: ModifiersState) -> MousePosition {
+        let mouse_mode =
+            TermMode::MOUSE_MOTION | TermMode::MOUSE_DRAG | TermMode::MOUSE_REPORT_CLICK;
+
         let buffer_point = self.ctx.terminal().visible_to_buffer(point);
 
         // Check message bar before URL to ignore URLs in the message bar
         if let Some(message) = self.message_at_point(Some(point)) {
             if self.message_close_at_point(point, message) {
-                MousePosition::MessageBarButton
+                return MousePosition::MessageBarButton;
             } else {
-                MousePosition::MessageBar
+                return MousePosition::MessageBar;
             }
-        } else if let Some(url) =
-            self.ctx.terminal().urls().drain(..).find(|url| url.contains(buffer_point))
+        }
+
+        // Check for URL at point with required modifiers held
+        if self.mouse_config.url.mods().relaxed_eq(modifiers)
+            && (!self.ctx.terminal().mode().intersects(mouse_mode) || modifiers.shift)
+            && self.mouse_config.url.launcher.is_some()
```

Comment:
I think url modifiers check can be moved to a separate function or back to `mouse_moved`, and we can just pass a `bool` (smth like search_for_urls) to mouse_position. This should make sense, since deciding to do smth is a `bool` and `ModifiersState`.

Reply:
The move to a separate function is motivating by this [problem](https://github.com/jwilm/alacritty/pull/2413#issuecomment-495264005)

Reply:
Hmm, I think I'd like to keep it here for now. It makes sense where it is and it's currently not needed anywhere else. Changing it also doesn't directly resolve the problem.

So I think this change is probably best done once a PR to resolve this is proposed.

Reply:
I just don't like that `mouse_position` requires now `ModifierState` as argument, so if we decided to reuse `mouse_position`, we'll deal with this restriction, which isn't generally a problem now, but can be annoying in some cases (using in `on_message_bar_click` in case we decide to add more buttons there or making `mouse_moved` public). So for now I have a strongly feeling that this code will be removed from this function anyway (and probably in a near feature), but it's up to you ofc.

Reply:
I don't think it's ever a good idea to write code for something that might or might not happen in the future. Even if there's a 100% chance that things are going to change in the future, the job of this PR is to write the best code possible *right now*. And I think that extracting that right now would just introduce additional complexity, rather than improving anything in its current situation.

If it's guaranteed to be changed in the future, that's fine. Let's do it in the future then.